src/types/sargon_types.rs

```diff
@@ -1111,3 +1563,254 @@ pub enum CommonError {
+    fn from(value: HierarchicalDeterministicFactorInstance) -> Self {
+        value.public_key_hash()
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, EnumAsInner)]
+pub enum ScryptoResourceOrNonFungible {
+    PublicKeyHash(PublicKeyHash),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, EnumAsInner)]
+pub enum ScryptoProofRule {
+    AnyOf(Vec<ScryptoResourceOrNonFungible>),
+    CountOf(usize, Vec<ScryptoResourceOrNonFungible>),
+    // AllOf
+    // Require
+    // AmountOf
+}
+impl ScryptoProofRule {
+    pub fn any_of(values: Vec<ScryptoResourceOrNonFungible>) -> Self {
+        Self::AnyOf(values)
+    }
+    pub fn count_of(count: usize, values: Vec<ScryptoResourceOrNonFungible>) -> Self {
+        Self::CountOf(count, values)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, EnumAsInner)]
+pub enum ScryptoAccessRuleNode {
+    ProofRule(ScryptoProofRule),
+    AnyOf(Vec<ScryptoAccessRuleNode>),
+    AllOf(Vec<ScryptoAccessRuleNode>),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, EnumAsInner)]
+pub enum ScryptoAccessRule {
+    Protected(ScryptoAccessRuleNode),
+    // AllowAll
+    // DenyAll
+}
+impl ScryptoAccessRule {
+    pub fn protected(rule: ScryptoAccessRuleNode) -> Self {
+        Self::Protected(rule)
+    }
+    pub fn with_threshold(
+        count: usize,
+        threshold_factors: impl IntoIterator<Item = impl Into<PublicKeyHash>>,
+        override_factors: impl IntoIterator<Item = impl Into<PublicKeyHash>>,
+    ) -> Self {
+        Self::protected(ScryptoAccessRuleNode::AnyOf(vec![
+            ScryptoAccessRuleNode::ProofRule(ScryptoProofRule::CountOf(
+                count,
+                threshold_factors
+                    .into_iter()
+                    .map(Into::into)
+                    .map(ScryptoResourceOrNonFungible::PublicKeyHash)
+                    .collect_vec(),
+            )),
+            ScryptoAccessRuleNode::ProofRule(ScryptoProofRule::AnyOf(
+                override_factors
+                    .into_iter()
+                    .map(Into::into)
+                    .map(ScryptoResourceOrNonFungible::PublicKeyHash)
+                    .collect_vec(),
+            )),
+        ]))
+    }
+}
+
+pub type MatrixOfKeyHashes = MatrixOfFactors<PublicKeyHash>;
+impl From<MatrixOfFactorInstances> for ScryptoAccessRule {
+    fn from(value: MatrixOfFactorInstances) -> Self {
+        Self::with_threshold(
+            value.threshold as usize,
+            value.threshold_factors,
+            value.override_factors,
+        )
+    }
+}
+impl From<MatrixOfKeyHashes> for ScryptoAccessRule {
+    fn from(value: MatrixOfKeyHashes) -> Self {
+        Self::with_threshold(
+            value.threshold as usize,
+            value.threshold_factors,
+            value.override_factors,
+        )
+    }
+}
+impl TryFrom<ScryptoAccessRule> for MatrixOfKeyHashes {
+    type Error = CommonError;
+
+    fn try_from(value: ScryptoAccessRule) -> Result<Self> {
```

Comment:
SHOULD only fail if a NON-Radix-Wallet client has implemented different ScryptoAccessRules - not adhering to our standard.